Utils/azuretools-core/src/com/microsoft/azuretools/ui/embeddedbrowser/JxBrowserUtil.java

```diff
@@ -22,42 +22,303 @@
+                FileUtils.deleteQuietly(filePathNames.get(i).toFile());
+                FileUtils.deleteQuietly(digestPathNames.get(i).toFile());
+            }
+
+            List<Callable<Boolean>> downloadThreads = new ArrayList<>();
+            for (i = 0; i < filePathNames.size(); i++) {
+                final Path filePathName = filePathNames.get(i);
+                final Path digestPathName = digestPathNames.get(i);
+                downloadThreads.add(new Callable<Boolean>() {
+                    @Override
+                    public Boolean call() throws Exception {
+                        return downloadFromAzure(filePathName);
+                    }
+                });
+
+                downloadThreads.add(new Callable<Boolean>() {
+                    @Override
+                    public Boolean call() throws Exception {
+                        return downloadFromAzure(digestPathName);
+                    }
+                });
+            };
+
+            List<Future<Boolean>> results = threadPool.invokeAll(downloadThreads, waitTime, TimeUnit.SECONDS);
+
+            for (i = 0; i < results.size(); i++) {
+                results.get(i).get();
+            }
+        } catch (Exception ex) {
+            result = false;
+            log.warning(String.format("Download fileName %s fail with exception %s", (filePathNames.size() > 0 ? filePathNames.get(0) : ""), ex.getMessage()));
+        } finally {
+            if (!result) {
+                log.warning("Start to download again with retry count " + currentRetry + " and wait time " + waitTime);
+                result = downloadAndCheckDigest(threadPool, filePathNames, currentRetry + 1, waitTime * RETRY_BACKOFF_FACTOR);
+            }
+        }
+
+        return result;
+    }
+
+    private static boolean downloadAndCheckDigest(@NotNull ExecutorService threadPool, @NotNull List<Path> filePathNames) {
+        return downloadAndCheckDigest(threadPool, filePathNames, 0, MAX_WAIT_TIME_IN_SECONDS);
+    }
+
+    private static String getJxBrowserJarFileName() throws JxBrowserException {
+        String osName = System.getProperty("os.name").toLowerCase();
+        String osArch = System.getProperty("os.arch").toLowerCase();
+        log.warning("osName is : " + osName + ", osArch is: " + osArch);
+
+        if (osName.contains("mac") || osName.contains("darwin")) {
+            return platformToJxBrowserJar.get("mac");
+        } else if (osName.contains("win")) {
+            if (osArch.contains("64")) {
+                return platformToJxBrowserJar.get("win64");
+            } else {
+                return platformToJxBrowserJar.get("win32");
+            }
+        } else if (osName.contains("nix") || osName.contains("nux")) {
+            if (osArch.contains("64")) {
+                return platformToJxBrowserJar.get("linux64");
+            } else {
+                return platformToJxBrowserJar.get("linux32");
```

Comment:
too many exits, suggest to assign the platform name into a var, and call `platformToJxBrowserJar.get()` at the function exit.

Reply:
Fixed.

---
In reply to: [156554588](https://github.com/Microsoft/azure-tools-for-java/pull/1296#discussion_r156554588) [](ancestors = 156554588)